File: Lab5/Python/Lecturer.py
from Person import *
class Lecturer(Person):
    def __init__(self, string):
        lst = string.split(";")
        super().__init__(lst[0], lst[1])
        self.__dis = {"disname": [], "hours": []}
        for i in range(2, len(lst), 2):
            d, h = lst[i:i + 2][0], lst[i:i + 2][1]
            self.__dis["disname"].append(d)
            self.__dis["hours"].append(int(h))

    # Підрахунок місячної заробітньої плати
    def earn(self):
        hours = 0
        for hour in self.__dis["hours"]:
            hours += hour
        return 100*hours

    # Повертає список дисциплін
    def GetDis(self):
        return self.__dis

    # Перевантаження виведення
    def __str__(self):
        res = f"{self._name};{self._date}"
        dis = self.__dis
        for i in range(len(dis["hours"])):
            res += ";{};{}".format(dis["disname"][i],dis["hours"][i])
        return res

File: Lab5/Lecturer.cpp
#include "Lecturer.h"

Lecturer::Lecturer()
{
    this->fname = "Noname";
    this->date = "01.01.1901";
}

Lecturer::Lecturer(std::string fname, std::string date, std::vector<std::string> disname, std::vector<int> hours)
{
    this->fname = fname;
    this->date = date;
    disciplines dat;
    for (int i = 0; i < disname.size(); i++)
    {
        dat.disname = disname[i];
        dat.hours = hours[i];
        this->dis.push_back(dat);
    }
}

Lecturer::Lecturer(std::string str)
{ 
    std::string fname, date;
    std::vector<std::string> disname;
    std::vector<int> hours;
    int n1 = 0, n2 = str.find(';');
    fname = str.substr(n1,n2-n1);
    n1 = n2+1;
    n2 = str.find(';', n1+1);
    date = str.substr(n1,n2-n1);
    n1 = n2+1;
    n2 = str.find(';', n1+1);
    std::vector<std::string> words;
    while(n2!=-1)
    {
        words.push_back(str.substr(n1,n2-n1));
        n1 = n2+1;
        n2 = str.find(';', n1+1);
    }
    words.push_back(str.substr(n1,str.size()-n1));
    for(int i = 0; i<words.size(); i+=2)
    {
        disname.push_back(words[i]);
    }
    for(int i = 1; i<words.size(); i+=2)
    {
        hours.push_back(stoi(words[i]));
    }    
    this->fname = fname;
    this->date = date;
    disciplines dat;
    for (int i = 0; i < disname.size(); i++)
    {
        dat.disname = disname[i];
        dat.hours = hours[i];
        this->dis.push_back(dat);
    }
} 

float Lecturer::earn()
{
    int hours = 0;
    for (int i = 0; i < this->dis.size(); i++)
    {
        hours += this->dis[i].hours;
    }
    return 100*hours;
}

std::vector<disciplines> Lecturer::GetDis()
{
    return this->dis;
}